File: CpuA64/disassembler_tool.py
#!/usr/bin/python3

import sys

from CpuA64 import symbolic
from CpuA64 import opcode_tab as a64


def disass(data):
    ins = a64.Disassemble(data)
    if ins is None:
        print(f"could not disassemble {data:x}")
        return

    enum_name, ops_str = symbolic.InsSymbolize(ins)
    print(f"{data:08x}", f"{ins.opcode.NameForEnum()} {' '.join(ops_str)}")
    for f, o, o_str in zip(ins.opcode.fields, ins.operands, ops_str):
        print(f"    {f.name:35s} {o_str:10} ({o})")
    # print(f"flags: {ins.opcode.classes}")
    print()
    data2 = a64.Assemble(ins)
    assert data == data2
    ins2 = symbolic.InsFromSymbolized(enum_name, ops_str)
    assert tuple(ins.operands) == tuple(
        ins2.operands), f"{ins.operands} vs {ins2.operands}"


def batch():
    for line in sys.stdin:
        if not line or line.startswith("#"): continue
        data = int(line.split()[0], 16)
        ins = a64.Disassemble(data)
        if ins.opcode is None:
            print(f"could not disassemble {data:x}")
            continue
        enum_name, ops_str = symbolic.InsSymbolize(ins)
        print(f"{data:08x} {enum_name}{' ' if ops_str else ''}{', '.join(ops_str)}")
        data2 = a64.Assemble(ins)
        assert data == data2
        ins2 = symbolic.InsFromSymbolized(enum_name, ops_str)
        assert tuple(ins.operands) == tuple(
            ins2.operands), f"{ins.operands} vs {ins2.operands}"


if __name__ == "__main__":
    if len(sys.argv) > 1:
        if sys.argv[1] == "batch":
            batch()
        else:
            for arg_hex_number in sys.argv[1:]:
                disass(int(arg_hex_number, 16))

File: CpuA64/disassembler_tool.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA64/symbolic.h"
#include "Util/assert.h"

#include <cstdlib>
#include <iostream>
#include <string_view>

using namespace cwerg::a64;

int main(int argc, char* argv[]) {
  if (argc <= 1) {
    std::cout << "no command specified\n";
    return 1;
  }
  if (std::string_view("disass") == argv[1]) {
    for (unsigned i = 2; i < argc; ++i) {
      const uint32_t data = strtoul(argv[i], nullptr, 16);
      Ins ins;
      if (!Disassemble(&ins, data)) {
        std::cout << "could not find opcode for: " << std::hex << data << "\n";
        continue;
      }
      std::vector<std::string> ops;
      std::string_view enum_name = InsSymbolize(ins, &ops);
      std::cout << argv[i] << " " << enum_name;
      std::string_view sep = " ";
      for (const std::string& op : ops) {
        std::cout << sep << op;
        sep = ", ";
      }
      std::cout << "\n";
      // check that the assembler works - this is not strictly
      // necessary but useful for debuggging the assembler
      const uint32_t data2 = Assemble(ins);
      if (data != data2) {
        std::cout << "Disassembler failure " << std::hex << data << " vs "
                  << data2 << "\n";
        return 1;
      }
      ASSERT(data == data2, "");
    }
  } else if (std::string_view("benchmark") == argv[1]) {
    // The first 1<<28 (256M) bit patterns exercise all opcodes with predicate
    // "eq"
    uint32_t num_bad = 0;
    std::vector<std::string> ops;
    for (uint32_t data = 0; data < (1U << 28U); ++data) {
      Ins ins;
      if (Disassemble(&ins, data)) {
        const uint32_t data2 = Assemble(ins);
        if (data != data2) {
          std::cout << "Disassembler failure " << std::hex << data << " vs "
                    << data2 << ": ";
          ops.clear();
          std::string_view enum_name = InsSymbolize(ins, &ops);
          std::cout << enum_name;
          std::string_view sep = " ";
          for (const std::string& op : ops) {
            std::cout << sep << op;
            sep = ", ";
          }
          std::cout << "\n";
          return 1;
        }
      } else {
        ++num_bad;
      }
      if (data % 10000000 == 0) {
        std::cout << num_bad << "/" << data << "\n";
      }
    }
    std::cout << "unsupported opcodes: " << std::dec << num_bad << "\n";
    return 0;
  } else {
    std::cout << "unknown command specified: " << argv[1] << "\n";
    return 1;
  }
}
